File: plugins/collector/event/apps/database/mongo.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"strings"
)

type MongoDB struct {
	version string
}

func (MongoDB) Name() string { return "mongo" }

func (MongoDB) Type() string { return "database" }

func (m MongoDB) Version() string { return m.version }

func (m *MongoDB) Match(p *process.Process) bool { return p.Name == "mongod" }

func (m *MongoDB) Run(p *process.Process) (mapping map[string]string, err error) {
	result, err := apps.Execute(p, "--version")
	if err != nil {
		return nil, err
	}
	m.version = strings.TrimPrefix(strings.Split(result, "\n")[0], "db version v")
	return
}

func init() { apps.Regist(&MongoDB{}) }

File: plugins/collector/src/event/apps/database/mongo.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct MongoDB {
    version: String,
}

impl MongoDB {
    pub fn new() -> Self { MongoDB { version: String::new() } }
}

impl IApp for MongoDB {
    fn name(&self)     -> &'static str { "mongo" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "mongod" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "db version v6.0.1"
        self.version = out.lines().next()
            .map(|l| l.trim_start_matches("db version v").trim().to_owned())
            .unwrap_or_default();
        Ok(HashMap::new())
    }
}
